codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/waiters/WaitableGenerator.kt

```diff
@@ -0,0 +1,258 @@
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.preludeScope
+import software.amazon.smithy.rust.codegen.core.smithy.RustCrate
+import software.amazon.smithy.rust.codegen.core.util.getTrait
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.core.util.outputShape
+import software.amazon.smithy.rust.codegen.core.util.toPascalCase
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+import software.amazon.smithy.waiters.WaitableTrait
+import software.amazon.smithy.waiters.Waiter
+
+class WaitableGenerator(
+    private val codegenContext: ClientCodegenContext,
+    allOperations: List<OperationShape>,
+) {
+    private data class WaitableOp(
+        val shape: OperationShape,
+        val waiters: List<WaiterSpec>,
+    )
+
+    private data class WaiterSpec(
+        val waiterName: String,
+        val waiter: Waiter,
+        val fluentBuilder: RuntimeType,
+    )
+
+    private val operations =
+        allOperations.mapNotNull { op ->
+            op.getTrait<WaitableTrait>()?.let {
+                WaitableOp(
+                    op,
+                    it.waiters.entries.map { (name, waiter) ->
+                        WaiterSpec(
+                            name,
+                            waiter,
+                            waiterFluentBuilder(name, waiter, op),
+                        )
+                    },
+                )
+            }
+        }
+            .sortedBy { it.shape.id }
+
+    fun render(crate: RustCrate) {
+        if (operations.isEmpty()) {
+            return
+        }
+
+        crate.withModule(ClientRustModule.client) {
+            docs(
+                """
+                Waiter functions for the client.
+
+                Import this trait to get `wait_for` methods on the client.
+                """,
+            )
+            rustBlockTemplate("pub trait Waiters") {
+                for (op in operations) {
+                    for (spec in op.waiters) {
+                        val waiterDocs = spec.waiter.documentation.orNull() ?: "Wait for `${spec.waiterName.toSnakeCase()}`"
+                        docs(waiterDocs)
+                        renderWaiterFnDeclaration(spec)
+                        rust(";")
+                    }
+                }
+            }
+
+            rustBlockTemplate("impl Waiters for Client") {
+                for (op in operations) {
+                    for (spec in op.waiters) {
+                        renderWaiterFnDeclaration(spec)
+                        rustTemplate(
+                            "{ #{FluentBuilder}::new(self.handle.clone()) }",
+                            "FluentBuilder" to spec.fluentBuilder,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    private fun RustWriter.renderWaiterFnDeclaration(spec: WaiterSpec) {
+        val fnName = "wait_for_${spec.waiterName.toSnakeCase()}"
```

Comment:
**discuss**

This name is fine I suppose (especially since there doesn't seem to be a recommended naming pattern) but I do wonder if it makes sense to align with other SDKs here.

It looks like [Go v1](https://github.com/aws/aws-sdk-go/blob/main/service/ec2/waiters.go#L16), [Java v2](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/waiters.html#id1waiters), [Swift](https://github.com/awslabs/aws-sdk-swift/blob/main/Sources/Services/AWSS3/Waiters.swift#L34), [Kotlin](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/waiters.html), [Ruby](https://docs.aws.amazon.com/sdk-for-ruby/v3/developer-guide/using-waiters.html), [PHP](https://docs.aws.amazon.com/en_us/aws-sdk-php/guide/latest/guide/waiters.html), and possibly others use `wait until` instead of `wait for`.  Of course there are outliers (Python and Go v2, didn't look at others). My vote would be to align with `wait_until` though.

Reply:
Ah. Good call. Thanks for looking at this. I'll rename them.